tests-and-examples/tests/cases/stream.rs

```diff
@@ -0,0 +1,148 @@
+        .bind("127.0.0.1", 0)
+        .build()
+        .unwrap();
+    server.start();
+    let port = server.bind_addrs().next().unwrap().1;
+    let ch = ChannelBuilder::new(env).connect(&format!("127.0.0.1:{}", port));
+    let client = RouteGuideClient::new(ch);
+
+    let pool = ThreadPoolBuilder::new().pool_size(2).create().unwrap();
+
+    let exec_test_f = async move {
+        // Test for send all disable batch
+        let (mut sink, receiver) = client.record_route().unwrap();
+        let mut send_data = vec![];
+        for i in 0..MESSAGE_NUM {
+            let mut p = Point::default();
+            p.set_longitude(i);
+            send_data.push(p);
+        }
+        let send_stream = futures::stream::iter(send_data);
+        sink.send_all(&mut send_stream.map(move |item| Ok((item, WriteFlags::default()))))
+            .await
+            .unwrap();
+        let summary = receiver.await.unwrap();
+        assert_eq!(summary.get_point_count(), MESSAGE_NUM);
+
+        // Test for send all enable batch
+        let (mut sink, receiver) = client.record_route().unwrap();
+        let mut send_data = vec![];
+        for i in 0..MESSAGE_NUM {
+            let mut p = Point::default();
+            p.set_longitude(i);
+            send_data.push(p);
+        }
+        let send_stream = futures::stream::iter(send_data);
+        sink.enhance_batch(true);
+        sink.send_all(&mut send_stream.map(move |item| Ok((item, WriteFlags::default()))))
+            .await
+            .unwrap();
+        let summary = receiver.await.unwrap();
+        assert_eq!(summary.get_point_count(), MESSAGE_NUM);
+
+        // Test for send all and all buffer hints are true
+        let (mut sink, receiver) = client.record_route().unwrap();
+        let mut send_data = vec![];
+        for i in 0..MESSAGE_NUM {
+            let mut p = Point::default();
+            p.set_longitude(i);
+            send_data.push(p);
+        }
+        let send_stream = futures::stream::iter(send_data);
+        sink.enhance_batch(false);
+        sink.send_all(
+            &mut send_stream.map(move |item| Ok((item, WriteFlags::default().buffer_hint(true)))),
+        )
+        .await
+        .unwrap();
+        // The following code is to test that when all msgs are set to be buffered, the msgs
+        // should be stored in the buffer until `sink.close()` is called.
+        let (tx, rx) = std::sync::mpsc::channel();
+        let close_sink_task = async move {
+            rx.recv_timeout(std::time::Duration::from_secs(1))
+                .unwrap_err();
+            sink.close().await.unwrap();
+            rx.recv_timeout(std::time::Duration::from_secs(1)).unwrap();
+        };
+        let recv_msg_task = async move {
+            let summary = receiver.await.unwrap();
+            tx.send(()).unwrap();
+            assert_eq!(summary.get_point_count(), MESSAGE_NUM);
+        };
+        join!(close_sink_task, recv_msg_task);
```

Comment:
The case is very hacky. It relies on scheduler polls future in FIFO order. A better approach is use futures channel instead of blocking.

Reply:
ou can adjust the sequence of close_sink_task and recv_msg_task and it doesn't matter. If use futures channel, this case cannot prove that all msgs will be buffered util sink.close() is called

Reply:
You can combine channel with timer to get `recv_timeout`.